scanners/scannerabc.py

```diff
@@ -0,0 +1,43 @@
+from abc import ABCMeta, abstractmethod
+from typing import Type, TypeVar
+import os
+
+ScannerABCT = TypeVar("ScannerABCT")
+
+
+class ScannerABC(metaclass=ABCMeta):
+
+    def __init__(self, domain: str, handles: dict, environment: dict,
+                 options: dict, extra: dict={}) -> None:
+        self.domain = domain
+        self.handles = handles
+        self.environment = {**environment, **extra}
+        self.options = options
+        self.report_dir = self.options.get("output", "./")
+        self.cache_dir = os.path.join(self.report_dir, "cache")
+        self.results_dir = os.path.join(self.report_dir, "results")
+
+        # Some subclasses may define an ``extra_environment`` property in their
+        # __init__ methods, and we want to fold its values into the
+        # ``environment`` property.
+        self.environment.update(getattr(self, "extra_environment", {}))
+
+    @abstractmethod
+    def scan(self) -> dict:
+        pass
+
+    @abstractmethod
+    def to_rows(self, data):
+        pass
+
+    @classmethod
+    @abstractmethod
+    def initialize_environment(cls: Type[ScannerABCT], environment: dict,
+                               options: dict) -> dict:
```

Comment:
I would prefer to see the call signature for scanners stay as it's been. I would also like to avoid passing `TypeVar`'s into individual scanners, with the goal of keeping the task of writing a scanner as simple and approachable as possible.

Reply:
I'm not sure why this is a factor; I'm assuming that we'll have a template for scanners that writers will copy from, meaning they won't have to worry about this type declaration. As for calling this class method, it remains the same, i.e. `scan_class.initialize_environment(scan_class, environment, options)`. Am I missing a consideration here?

Reply:
> I'm assuming that we'll have a template for scanners that writers will copy from, meaning they won't have to worry about this type declaration.

I'm still assuming scanners will be written by hand, and would like that to be as accessible as possible. But even when using a template, I'd like it to be clear what's going on in the file, and what each function means and does.

Reply:
> As for calling this class method, it remains the same, i.e. scan_class.initialize_environment(scan_class, environment, options). Am I missing a consideration here?

I think so? There's never been an `initialize_environment(scan_class, environment, options)` function. Only an `init(environment, options)` function.